### PR TITLE
[webpack-assets-manifest]: Add missing types, fix existing types & update to latest minor versions

### PR DESCRIPTION
**Pull request by kira-bruneau**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  **Overview of changes in v4:**
  - https://github.com/webdeveric/webpack-assets-manifest/tree/v5.0.6#new-in-version-4

  **Overview of changes in v5:**
  - https://github.com/webdeveric/webpack-assets-manifest/tree/v5.0.6#new-in-version-5

  **`WebpackAssetsManifest.Options` type derived from:**
  - https://github.com/webdeveric/webpack-assets-manifest/blob/v5.0.6/src/options-schema.json

  **`WebpackAssetsManifest.Options.replacer` type derived from:**
  - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_replacer_parameter
  - https://github.com/microsoft/TypeScript/blob/v4.3.5/lib/lib.es5.d.ts#L1066
  - https://github.com/microsoft/TypeScript/blob/v4.3.5/lib/lib.es5.d.ts#L1073

  **`WebpackAssetsManifest` type derived from:**
  - https://github.com/webdeveric/webpack-assets-manifest/blob/v5.0.6/src/WebpackAssetsManifest.js
- [x] ~~If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~~

### CODE REVIEW

File: types/webpack-assets-manifest/v4/index.d.ts [hunk 1]
```diff
@@ -33,66 +34,145 @@ declare class WebpackAssetsManifest extends Plugin {
...
-        replacer?: null | ReadonlyArray<string> | ((key: string, value: string) => number | string | boolean | null | object | undefined) | undefined;
+        replacer?:
+            | ((this: unknown, key: string, value: unknown) => unknown)
+            | ReadonlyArray<string | number>
+            | null
+            | undefined;
 
         /** https://github.com/webdeveric/webpack-assets-manifest#space */
-        space?: number | undefined;
+        space?: number | string | undefined;
```
Comment (FranklinWhale) [inline types/webpack-assets-manifest/v4/index.d.ts:172]: Same as v5. I think string should not be allowed.

File: types/webpack-assets-manifest/index.d.ts [hunk 2]
```diff
@@ -5,98 +5,183 @@
...
-        replacer?: null | ReadonlyArray<string> | ((key: string, value: string) => number | string | boolean | null | object | undefined) | undefined;
+        replacer?:
+            | ((this: unknown, key: string, value: unknown) => unknown)
+            | ReadonlyArray<string | number>
+            | null
+            | undefined;
 
         /** https://github.com/webdeveric/webpack-assets-manifest#space */
-        space?: number | undefined;
+        space?: number | string | undefined;
```
Comment (FranklinWhale) [inline types/webpack-assets-manifest/index.d.ts:181]: The doc says it is `int`. I think `string` should not be allowed.

**Review by FranklinWhale — COMMENTED** [review]
The change from `any` to `unknown` is a breaking change. I am not sure whether the use of `unknown` is common in `.d.ts`.

**Comment by kira-bruneau** [comment]
> The change from `any` to `unknown` is a breaking change. I am not sure whether the use of `unknown` is common in `.d.ts`.

I was just following the instructions under [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes). Specifically https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any.

I originally had them typed as `any`before following those instructions though, and I could just revert the changes.

File: types/webpack-assets-manifest/index.d.ts [hunk 3]
```diff
@@ -5,98 +5,183 @@
...
     /**
-     * Get a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler) for the manifest
+     * Get a {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler|Proxy} for the manifest.
+     * This allows you to use `[]` to manage entries.
      *
-     * @param raw - Use `setRaw` instead of `set`
+     * @param raw - Should the proxy use `setRaw` instead of `set`?
      */
-    getProxy(raw?: boolean): ProxyHandler<WebpackAssetsManifest>;
+    getProxy(raw?: boolean): WebpackAssetsManifest.Assets;
```
Comment (kira-bruneau) [inline types/webpack-assets-manifest/index.d.ts:159]: `ProxyHandler` was the wrong type, but I'm not sure if this is exactly the right type either. The way the proxy is set up, enumerating over the keys of this value will enumerate over the keys of `WebpackAssetsManifest`.

**Comment by FranklinWhale** [comment]
> > The change from `any` to `unknown` is a breaking change. I am not sure whether the use of `unknown` is common in `.d.ts`.
> 
> I was just following the instructions under [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes). Specifically https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any.
> 
> I originally had them typed as `any` before following those instructions though, and I could just revert the changes.

Thanks for the reference. I think it's fine to use `unknown`.

File: types/webpack-assets-manifest/index.d.ts (hunk 3 above)
Comment (FranklinWhale) [reply types/webpack-assets-manifest/index.d.ts:159]: I think it's fine too

**Review by FranklinWhale — APPROVED** [review]
LGTM. Thanks!

**Comment by kira-bruneau** [comment]
Ready to merge

**Comment by damiangreen** [comment]
@kira-bruneau  I am getting `Excessive stack depth comparing types '{ name: string; target: string; entry: { client: string[]; }; plugins: (WebpackAssetsManifest | ((this: Compiler, compiler: Compiler) => void) | WebpackPluginInstance)[]; optimization: { splitChunks: { chunks: "all"; }; }; amd?: false | { [index: string]: any; }; bail?: boolean; cache?: boolean | MemoryCacheOptions | FileCacheOptions; context?: string; dependencies?: string[]; devtool?: string | false; experiments?: Experiments; externals?: string | RegExp | (ExternalItemObjectKnown & ExternalItemObjectUnknown) | ExternalItem[] | ((data: ExternalItemFunctionData, callback: (err?: Error, result?: string | boolean | string[] | { [index: string]: any; }) => void) => void) | ((data: ExternalItemFunctionData) => Promise<ExternalItemValue>); externalsPresets?: ExternalsPresets; externalsType?: "global" | "module" | "import" | "var" | "assign" | "this" | "window" | "self" | "commonjs" | "commonjs2" | "commonjs-module" | "amd" | "amd-require" | "umd" | "umd2" | "jsonp" | "system" | "promise" | "script" | "node-commonjs"; ignoreWarnings?: (RegExp | { file?: RegExp; message?: RegExp; module?: RegExp; } | ((warning: WebpackError, compilation: Compilation) => boolean))[]; infrastructureLogging?: InfrastructureLogging; loader?: Loader; mode?: "development" | "production" | "none"; module?: ModuleOptions; node?: false | NodeOptions; output?: Output; parallelism?: number; performance?: false | PerformanceOptions; profile?: boolean; recordsInputPath?: string | false; recordsOutputPath?: string | false; recordsPath?: string | false; resolve?: ResolveOptionsWebpackOptions; resolveLoader?: ResolveOptionsWebpackOptions; snapshot?: SnapshotOptions; stats?: boolean | "none" | "verbose" | "summary" | "errors-only" | "errors-warnings" | "minimal" | "normal" | "detailed" | StatsOptions; watch?: boolean; watchOptions?: WatchOptions; devServer?: Configuration; }' and 'Configuration'`  when using these latest typings...

**Pull request merged**